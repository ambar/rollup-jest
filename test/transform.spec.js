import inject from '@rollup/plugin-inject'
import nodeResolve from '@rollup/plugin-node-resolve'
import {transform} from '..'
import {resolve} from 'path'

describe('process', () => {
  const warn = jest.fn(console.warn)
  const oldConsole = global.console
  global.console = {...oldConsole, warn}
  afterAll(() => {
    global.console = oldConsole
  })

  it('should transform', async () => {
    const code = `
        export const foo =  42
        export {default as noop} from 'noop3'
        export {URL} from 'url'
    `
    const file = './null.js'
    expect(await transform({code, file})).toMatchSnapshot()
    expect(warn.mock.calls.join('')).not.toMatch(
      /could not be resolved – treating it as an external dependency/
    )
  })

  it('should provide a sourcemap if requested', async () => {
    const code = `
        export const foo =  42
        export {default as noop} from 'noop3'
        export {URL} from 'url'
    `
    const file = './null.js'
    expect(
      await transform({code, file}, {output: {sourcemap: true}})
    ).toMatchSnapshot()
    expect(warn.mock.calls.join('')).not.toMatch(
      /could not be resolved – treating it as an external dependency/
    )
  })

  it('should transform imports if requested', async () => {
    const code = `
      import { foo } from './fixtures/utils'
      export {URL} from 'url'

      console.log(foo)
    `
    const file = resolve(__dirname, './null.js')
    expect(
      await transform({code, file}, {resolveImports: true})
    ).toMatchSnapshot()
    expect(warn.mock.calls.join('')).not.toMatch(
      /could not be resolved – treating it as an external dependency/
    )
  })

  it('should transform imports if requested, with nodeResolve', async () => {
    // for proper code, we'd also need commonjs to properly resolve noop3 (CJS module), but it's enough for the test
    const code = `
      import * as noop from 'noop3'
      import { foo } from './fixtures/utils'
      export {URL} from 'url'

      console.log(foo, noop)
    `
    const file = resolve(__dirname, './null.js')
    expect(
      await transform(
        {code, file},
        {resolveImports: true, plugins: [nodeResolve()]}
      )
    ).toMatchSnapshot()
    expect(warn.mock.calls.join('')).not.toMatch(
      /could not be resolved – treating it as an external dependency/
    )
  })

  it('should transform imports if requested relative', async () => {
    const code = `
      import * as noop from 'noop3'
      import { foo } from './fixtures/utils'
      export {URL} from 'url'

      console.log(foo, noop)
    `
    const file = resolve(__dirname, './null.js')
    expect(
      await transform({code, file}, {resolveImports: 'relative'})
    ).toMatchSnapshot()
    expect(warn.mock.calls.join('')).not.toMatch(
      /could not be resolved – treating it as an external dependency/
    )
  })

  it('should add custom plugin', async () => {
    const code = `noop()`
    const file = './null.js'
    expect(
      await transform(
        {code, file},
        {
          plugins: [inject({noop: 'noop3'})],
        }
      )
    ).toMatch(/noop3/)
  })

  it('should add custom plugin with babel-like require', async () => {
    const code = `noop()`
    const file = './null.js'
    expect(
      await transform(
        {code, file},
        {
          plugins: [['@rollup/plugin-inject', {noop: 'noop3'}]],
        }
      )
    ).toMatch(/noop3/)
  })

  it('should add custom plugin with cjs config', async () => {
    const code = `noop()`
    const file = './null.js'
    expect(
      await transform(
        {code, file},
        {
          configFile: 'test/fixtures/config.cjs.js',
        }
      )
    ).toMatch(/noop3/)
  })

  it('should add custom plugin with esm config', async () => {
    const code = `noop()`
    const file = './null.js'
    expect(
      await transform(
        {code, file},
        {
          configFile: 'test/fixtures/config.esm.js',
        }
      )
    ).toMatch(/noop3/)
  })

  it('should add custom plugin with cjs config function', async () => {
    const code = `noop()`
    const file = './null.js'
    expect(
      await transform(
        {code, file},
        {
          configFile: 'test/fixtures/config.cjs.js',
          args: {noop: 'noop3'},
        }
      )
    ).toMatch(/noop3/)
  })

  it('should add custom plugin with esm config that imports esm config', async () => {
    const code = `noop()`
    const file = './null.js'
    expect(
      await transform(
        {code, file},
        {
          configFile: 'test/fixtures/config.import.js',
        }
      )
    ).toMatch(/noop3/)
  })
})
