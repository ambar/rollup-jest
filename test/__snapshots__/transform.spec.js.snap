// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process should provide a sourcemap if requested 1`] = `
Object {
  "code": "'use strict';

var noop3 = require('noop3');
var url = require('url');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var noop3__default = /*#__PURE__*/_interopDefault(noop3);

const foo =  42;

Object.defineProperty(exports, 'noop', {
    enumerable: true,
    get: function () { return noop3__default.default; }
});
Object.defineProperty(exports, 'URL', {
    enumerable: true,
    get: function () { return url.URL; }
});
exports.foo = foo;
//# sourceMappingURL=null.js.map
",
  "map": SourceMap {
    "file": "null.js",
    "mappings": ";;;;;;;;;AACoB,MAAC,GAAG,IAAI;;;;;;;;;;;;",
    "names": Array [],
    "sources": Array [
      "null.js",
    ],
    "sourcesContent": Array [
      "
        export const foo =  42
        export {default as noop} from 'noop3'
        export {URL} from 'url'
    ",
    ],
    "version": 3,
  },
}
`;

exports[`process should transform 1`] = `
Object {
  "code": "'use strict';

var noop3 = require('noop3');
var url = require('url');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var noop3__default = /*#__PURE__*/_interopDefault(noop3);

const foo =  42;

Object.defineProperty(exports, 'noop', {
    enumerable: true,
    get: function () { return noop3__default.default; }
});
Object.defineProperty(exports, 'URL', {
    enumerable: true,
    get: function () { return url.URL; }
});
exports.foo = foo;
//# sourceMappingURL=null.js.map
",
  "map": SourceMap {
    "file": "null.js",
    "mappings": ";;;;;;;;;AACoB,MAAC,GAAG,IAAI;;;;;;;;;;;;",
    "names": Array [],
    "sources": Array [
      "null.js",
    ],
    "sourcesContent": Array [
      "
        export const foo =  42
        export {default as noop} from 'noop3'
        export {URL} from 'url'
    ",
    ],
    "version": 3,
  },
}
`;

exports[`process should transform imports if requested 1`] = `
Object {
  "code": "'use strict';

var url = require('url');

const foo = 42;

console.log(foo);

Object.defineProperty(exports, 'URL', {
    enumerable: true,
    get: function () { return url.URL; }
});
//# sourceMappingURL=null.js.map
",
  "map": SourceMap {
    "file": "null.js",
    "mappings": ";;;;AAAO,MAAM,GAAG,GAAG;;ACIb,OAAO,CAAC,GAAG,CAAC,GAAG;;;;;;;",
    "names": Array [],
    "sources": Array [
      "fixtures/utils.js",
      "null.js",
    ],
    "sourcesContent": Array [
      "export const foo = 42
",
      "
      import { foo } from './fixtures/utils'
      export {URL} from 'url'

      console.log(foo)
    ",
    ],
    "version": 3,
  },
}
`;

exports[`process should transform imports if requested relative 1`] = `
Object {
  "code": "'use strict';

var noop = require('noop3');
var url = require('url');

function _interopNamespace(e) {
    if (e && e.__esModule) return e;
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var noop__namespace = /*#__PURE__*/_interopNamespace(noop);

const foo = 42;

console.log(foo, noop__namespace);

Object.defineProperty(exports, 'URL', {
    enumerable: true,
    get: function () { return url.URL; }
});
//# sourceMappingURL=null.js.map
",
  "map": SourceMap {
    "file": "null.js",
    "mappings": ";;;;;;;;;;;;;;;;;;;;;;;;;AAAO,MAAM,GAAG,GAAG;;ACKb,OAAO,CAAC,GAAG,CAAC,GAAG,EAAEA,eAAI;;;;;;;",
    "names": Array [
      "noop",
    ],
    "sources": Array [
      "fixtures/utils.js",
      "null.js",
    ],
    "sourcesContent": Array [
      "export const foo = 42
",
      "
      import * as noop from 'noop3'
      import { foo } from './fixtures/utils'
      export {URL} from 'url'

      console.log(foo, noop)
    ",
    ],
    "version": 3,
  },
}
`;

exports[`process should transform imports if requested, with nodeResolve 1`] = `
Object {
  "code": "'use strict';

var url = require('url');

const noopFactory = require('./factory');

module.exports = noopFactory();

var noop = /*#__PURE__*/Object.freeze({
    __proto__: null
});

const foo = 42;

console.log(foo, noop);

Object.defineProperty(exports, 'URL', {
    enumerable: true,
    get: function () { return url.URL; }
});
//# sourceMappingURL=null.js.map
",
  "map": SourceMap {
    "file": "null.js",
    "mappings": ";;;;AACA,MAAM,WAAW,GAAG,OAAO,CAAC,WAAW,CAAC,CAAC;AACzC;AACA,MAAM,CAAC,OAAO,GAAG,WAAW,EAAE;;;;;;ACHvB,MAAM,GAAG,GAAG;;ACKb,OAAO,CAAC,GAAG,CAAC,GAAG,EAAE,IAAI;;;;;;;",
    "names": Array [],
    "sources": Array [
      "../node_modules/noop3/index.js",
      "fixtures/utils.js",
      "null.js",
    ],
    "sourcesContent": Array [
      "'use strict';
const noopFactory = require('./factory');

module.exports = noopFactory();
",
      "export const foo = 42
",
      "
      import * as noop from 'noop3'
      import { foo } from './fixtures/utils'
      export {URL} from 'url'

      console.log(foo, noop)
    ",
    ],
    "version": 3,
  },
}
`;
